### PR TITLE
IGNITE-17274 Improve service name lookup performance

### PR DESCRIPTION
**Pull request by artnaseef**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by artnaseef** [comment]
The change adds a Map of services by service-name in addition to the Map by service-id, and uses this map for lookups instead of a linear search of the map by service-name.

This significantly reduces the startup timing of services and changes the growth in timing from exponential to linear.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java [hunk 1]
```diff
@@ -1858,6 +1862,9 @@ private void processDynamicCacheChangeRequest(DynamicCacheChangeBatch msg) {
             depActions.servicesToUndeploy(toUndeploy);
 
             msg.servicesDeploymentActions(depActions);
+
+            // Remove the names from the service-by-name map as well
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1866]: I don't think this is correct. We should only update `registeredServicesByName` right where `registeredServices` is updated, so they are always in sync.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java [hunk 2]
```diff
@@ -168,6 +168,7 @@ public class IgniteServiceProcessor extends GridProcessorAdapter implements Igni
      * @see ServiceDeploymentManager
      */
     private final ConcurrentMap<IgniteUuid, ServiceInfo> registeredServices = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, ServiceInfo> registeredServicesByName = new ConcurrentHashMap<>();
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:171]: Missing javadoc (build will fail).
Let's mention that this map simply improves lookup by name and should be always in sync with the one above.

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java (hunk 1 above)
Comment (artnaseef) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1866]: Like this?

                registeredServices.entrySet().removeIf(e -> {
                    ServiceInfo desc = e.getValue();

                    if (Objects.equals(desc.cacheName(), chReq.cacheName())) {
                        toUndeploy.put(desc.serviceId(), desc);
                        registeredServicesByName.remove(desc.name());

                        return true;
                    }

                    return false;
                });

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java (hunk 2 above)
Comment (artnaseef) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:171]: ok, will do

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1866]: I mean, wherever we have `registeredServices.put` or `.remove` or `.clear`, we also update `registeredServicesByName`.

On this line there is no such thing, so let's revert it. Other cases are already handled.

Comment (artnaseef) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1866]: See the removeIf() line - it's a foreach + remove in one.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java (hunk 2 above)
Comment (artnaseef) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:176]: Javadoc has been added.

**Pull request merged**